movai_core_shared/core/zmq_server.py

```diff
@@ -0,0 +1,166 @@
+        self._socket.bind(self._addr)
+        self._logger.info(f"{self.__class__.__name__} is listening on {self._addr}")
+
+    async def _accept(self) -> None:
+        """accepts new connections requests to zmq."""
+        while self._running:
+            try:
+                if self._debug:
+                    self._logger.debug("Waiting for new requests.\n")
+                request = await self._socket.recv_multipart()
+                asyncio.create_task(self._handle(request))
+            except Exception as error:
+                self._logger.error(f"ZMQServer Error: {str(error)}")
+                continue
+
+    async def _handle(self, msg_buffer) -> None:
+        index = len(msg_buffer) - 1
+        buffer = msg_buffer[index]
+
+        if buffer is None:
+            self._logger.warning("Request has no buffer, can't handle request.")
+            return
+
+        request_msg = json.loads(buffer)
+        if "request" not in request_msg:
+            raise UnknownRequestError(f"The message format is unknown: {request_msg}.")
+        request = request_msg.get("request")
+        
+        if self._debug:
+            general_request = Request(**request)
+            self._logger.debug(general_request.__str__())
+        
+        response = await self.handle_request(request)
+        response_required = request.get("response_required")
+        if response_required:
+            if response.get("response") is None:
+                response = {"response": response}
+            response = await self.handle_response(response)
+            index = len(msg_buffer) - 1
+            msg_buffer[index] = json.dumps(response).encode("utf8")
+            self._socket.send_multipart(msg_buffer)
+            if self._debug:
+                self._logger.debug(f"{self._name} successfully sent a respone.")
+                
+
+
+    def _close(self) -> None:
+        """close the zmq socket."""
+        self._socket.close()
+        self._ctx.destroy()
+        self._socket = None
+        self._ctx = None
+        self._initialized = False
+
+    def __del__(self):
+        """closes the socket when the object is destroyed.
+        """
+        self._close()
+
+    def init_server(self):
+        """Initializes the server to listen on the specified address.
+        """
+        if self._initialized:
+            self._logger.error("Server is already initialized.")
+            return
+        
+        try:
+            self._set_context()
+        except OSError:
+            self._logger.error("failed to create server socket")
+            return 1
+
+        try:
+            self._bind()
+            self._initialized = True
+        except OSError as error:
+            self._logger.error(f"failed to bind socket on address {self._addr}")
+            raise
+
+    def run(self) -> int:
+        """The main message dispatch loop.
+
+        Returns:
+            int: 1 in case of exception.
+        """
+        self.init_server()
```

Comment:
if you created _initialized use it  
```python
if self._initialized or self._running:
    self._logger.warning('can't ask to run a server that runs')
```
or run it again

Reply:
the initialized flag is being checked inside init_server(). 
I did added a check for the self._running flag.